### PR TITLE
effects: annotate `nameof(::Module)` + `parentmodule(::Module)` for constant folding

### PR DESCRIPTION
**Pull request by topolarity**
These effects can be important for us to statically resolve Module-handling logic that, e.g., inspects a loaded list of modules looking for a particular name.

Co-authored-by: @gbaraldi

### CODE REVIEW

File: base/essentials.jl [hunk 1]
```diff
@@ -180,22 +180,11 @@ macro isdefined(s::Symbol)
...
-:Broadcast
-```
-"""
-nameof(m::Module) = ccall(:jl_module_name, Ref{Symbol}, (Any,), m)
+_nameof(m::Module) = ccall(:jl_module_name, Ref{Symbol}, (Any,), m)
 
 function _is_internal(__module__)
     if ccall(:jl_base_relative_to, Any, (Any,), __module__)::Module === Core.Compiler ||
-       nameof(__module__) === :Base
+       _nameof(__module__) === :Base
```
Comment (aviatesk) [inline base/essentials.jl:187]: Is this change necessary for some reason?

**Review by aviatesk — APPROVED** [review]

File: base/essentials.jl (hunk 1 above)
Comment (topolarity) [reply base/essentials.jl:187]: Yeah, this is used in the definition of the macro that is used to annotate effects in `nameof`

Comment (aviatesk) [reply base/essentials.jl:187]: Ah I see 😅

**Review by Keno — APPROVED** [review]
